### PR TITLE
CDAP-16268: Added generated fields to summaries

### PR DESCRIPTION
**Pull request by vravish**

### CODE REVIEW

File: cdap-data-fabric/src/test/java/io/cdap/cdap/data2/metadata/lineage/field/FieldLineageInfoTest.java [hunk 1]
```diff
@@ -41,6 +41,8 @@
 import java.util.Map;
 import java.util.Set;
 
+import static io.cdap.cdap.data2.metadata.lineage.field.FieldLineageInfo.NULL_EPF;
```
Comment (chtyim) [inline cdap-data-fabric/src/test/java/io/cdap/cdap/data2/metadata/lineage/field/FieldLineageInfoTest.java]: Don't use import static.

File: cdap-data-fabric/src/test/java/io/cdap/cdap/data2/metadata/lineage/field/FieldLineageInfoTest.java [hunk 2]
```diff
@@ -874,6 +880,92 @@ public void testMultiSourceDroppedFields() {
...
   }
 
+  @Test
+  public void testGeneratedFields() {
+    // read: endpoint1 -> (first_name, last_name)
+    // generateSocialAndDob: () -> social, dob
+    // renameSocial: generateSocialAndDob.social -> ssn
+    // write: (read.first_name, read.last_name, renameSocial.ssn, generateSocialAndDob.dob) -> endpoint2
+    ReadOperation read = new ReadOperation("read", "some read", EndPoint.of("endpoint1"),
+        "first_name", "last_name");
```
Comment (chtyim) [inline cdap-data-fabric/src/test/java/io/cdap/cdap/data2/metadata/lineage/field/FieldLineageInfoTest.java]: The alignment in the current code base is aligned to the method open `(`

```java
    ReadOperation read = new ReadOperation("read", "some read", EndPoint.of("endpoint1"),
                                           "first_name", "last_name");
```
```

File: cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 3]
```diff
@@ -376,12 +379,17 @@ private void computeIncomingSummaryHelper(EndPointField field, Operation current
...
+        ReadOperation read = (ReadOperation) currentOperation;
+        EndPoint source = read.getSource();
+        for (InputField inputField : inputFields) {
+          if (inputField.getOrigin().equals(currentOperation.getName())) {
+            sourceEndPointFields.add(new EndPointField(source, inputField.getName()));
+          }
         }
+      } else {
+        // note the source of every generated field is NULL_EPF
+        sourceEndPointFields.add(NULL_EPF);
```
Comment (yaojiefeng) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/lineage/field/FieldLineageInfo.java:392]: The logic here is mostly correct. There is only one problem: if there is a mixture of dropped and generated fields, the NULL_EPF will get removed at https://github.com/cdapio/cdap/blob/38c258f70ad937de67a44030719c6a7554b3c362/cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/lineage/field/FieldLineageInfo.java#L146. And the outgoing summary will have NULL_EPF which we don't need. We might need to change the key of the incomingSummary to an object that can identify the type of the operation.

**Review by yaojiefeng — COMMENTED** [review]
One comment about the mixture of dropped and generated field. Also please fix Terence's comments, thanks

File: cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 3 above)
Comment (vravish) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/lineage/field/FieldLineageInfo.java:392]: Can we fix this by not removing NULL_EPF if there are generatedFields?

**Pull request closed without merging**